proposals/2025-04-04_extended-range-selectors-semantics.md

```diff
@@ -0,0 +1,396 @@
+
+This is the result of calling `increase()` over the range:
+
+![](../assets/2025-04-04_extended-range-selectors-semantics/84f19b21166f516df4586326851ce8bf.png)
+
+When working with the **partial** dataset, Prometheus is interpolating only by half a scrape interval, which provides this:
+
+![](../assets/2025-04-04_extended-range-selectors-semantics/36b0b5c4d1261bcbba3e923b1adfbae7.png)
+
+We can clearly see that Prometheus is providing an increase of 2.17 for the first vector, which is an approximation based on the available data points within the range.
+
+When these same interpolation and extrapolation heuristics are applied consistently across the entire partial dataset, we get the following visualization:
+
+![](../assets/2025-04-04_extended-range-selectors-semantics/881987961c011e6f42321e778db91b5c.png)
+
+This demonstrates the function's poor behavior in non-ideal scenarios. The graph shows a significant spike that appears and vanishes immediately, highlighting the instability of the calculation when working with incomplete data.
+
+## Goals
+
+The core aim of this proposal is to offer greater composability and flexibility
+in how range selectors behave, especially for foundational functions like `rate`
+and `increase`. Instead of enforcing a single rigid interpretation of boundary
+semantics and extrapolation, the proposal adopts a **"choose what fits best"**
+philosophy—recognizing that different use cases benefit from different
+trade-offs.
+
+## Non-Goals
+
+While the proposal aims to improve expressiveness and correctness in range-based
+functions, it deliberately avoids certain directions that have historically led
+to complexity or confusion:
+
+- **Define a one-size-fits-all solution**: The goal is not to find a universally
+  "better" rate function that replaces all others. Different use cases have
+  inherently different trade-offs, and the proposal embraces offering multiple
+  well-defined modes instead of forcing consensus around a single behavior.
+- **Introduce new function names that encode behavior**: Rather than proliferating new
+  variants like xrate, yrate, arate, etc., this proposal keeps the function names
+  unchanged and introduces boundary behavior via modifiers. This preserves the
+  core vocabulary of PromQL and avoids unnecessary fragmentation.
+- **Make silent or implicit changes to existing functions**: Backward compatibility
+  is critical. The proposal does not seek to silently change the
+  semantics of rate, increase, or delta as they exist today. All new behaviors
+  must be opt-in and explicit.
+- **Abstract away all underlying heuristics**: This is not about hiding complexity—it’s
+  about making trade-offs more explicit and controllable. Users still need to
+  understand how their choice of modifier affects the results, and documentation
+  and naming should reflect that.
+
+## How
+
+### How This Proposal Works
+
+This proposal introduces two new **modifiers for range and instant selectors**
```

Comment:
Could you elaborate on why the new modifiers are necessary for instant vector selectors? It seems from the motivation above and discussion below the new modifiers are only necessary for range vector selectors.

Reply:
Makes sense. Perhaps we can start with adding these keywords just for range vector selectors for now?